tree/tree/src/TTree.cxx

```diff
@@ -4703,8 +4708,10 @@ Int_t TTree::Fill()
    // to the case where the tree is in the top level directory.
    if (fDirectory)
       if (TFile *file = fDirectory->GetFile())
-         if ((TDirectory *)file == fDirectory && (file->GetEND() > fgMaxTreeSize))
-            ChangeFile(file);
+         if (static_cast<TDirectory *>(file) == fDirectory && (file->GetEND() > fgMaxTreeSize))
+            // Changing file clashes with the design of TMemFile and derivates, see #6523.
+            if (!(dynamic_cast<TMemFile *>(file)))
+               ChangeFile(file);
```

Comment:
@pcanal, instead of checking for a `TMemFile` here, would there be a way to check for a file on disk and revert the condition? For example, if `file` is a `TDirectoryFile`. The question is really, for which of the classes below it makes sense to call `ChangeFile()`?
```bash
$ git grep -o 'class T.*File.*public T.*File'
dcache/inc/TDCacheFile.h:class TDCacheFile : public TFile
gfal/inc/TGFALFile.h:class TGFALFile : public TFile
io/inc/ROOT/TBufferMerger.hxx:class TBufferMergerFile : public TMemFile
io/inc/TFile.h:class TFile : public TDirectoryFile
io/inc/TMemFile.h:class TMemFile : public TFile
io/inc/TZIPFile.h:class TZIPFile : public TArchiveFile
mpi/inc/TMPIFile.h:class TMPIFile : public TMemFile
sql/inc/TSQLFile.h:class TSQLFile final : public TFile
xml/inc/TXMLFile.h:class TXMLFile final : public TFile
```

Reply:
The original issue and the new one described in #6640 are addressed in the PR, if you think the solution should be modified I'm open to discussion :smile: .